feat(mutation exclusion): mutants with ExcludeFromCodeCoverage attribute

riezebosch opened the pull request:
If you choose to exclude something from coverage I think it makes sense to also exclude it from mutation testing.

riezebosch commented on src/Stryker.Core/Stryker.Core/MutantFilters/ExcludeFromCodeCoverageFilter.cs:
```diff
@@ -0,0 +1,38 @@
+
+        private static bool HasExcludeFromCodeCoverageAttribute(MemberDeclarationSyntax m) =>
+            m.AttributeLists.Any(attr =>
+                attr.Attributes.Any(a => Regex.IsMatch(a.ToFullString(), @"^(?:System\.Diagnostics\.CodeAnalysis\.)?ExcludeFromCodeCoverage(?:Attribute)?$")));
```
Not sure if there's a better way to validate attributes without to much context.

richardwerkman replied:
I don't think there is. We won't calculate the semantic model for each file because of the performance costs. So we have no way of getting the full name of the attribute for now.

riezebosch replied:
By that I changed it to `ToString`

riezebosch commented:
Have ran it on my own realistic project and the recursive traversal of the tree handles this for sure but I agree on improving the mutants used in test.

riezebosch commented:
Nice: 
<pre><code>
[11:41:16 INF] Using testrunner VsTest
[11:41:16 INF] Total number of tests found: 230
[11:41:16 INF] Initial testrun started
[11:43:26 INF] Using 199299 ms as testrun timeout
[11:43:30 INF] <b>69 mutants got status Skipped. Reason: Removed by exclude from code coverage filter</b>
[11:43:30 INF] 78 mutants got status Skipped. Reason: Removed by method filter
[11:43:30 INF] 477 mutants ready for test
</code></pre>

richardwerkman commented on src/Stryker.Core/Stryker.Core/MutantFilters/ExcludeFromCodeCoverageFilter.cs:
```diff
@@ -0,0 +1,38 @@
+    {
+        string IMutantFilter.DisplayName { get; } = "exclude from code coverage filter";
+
+        IEnumerable<Mutant> IMutantFilter.FilterMutants(IEnumerable<Mutant> mutants, FileLeaf file, StrykerOptions options) =>
```
```suggestion
        IEnumerable<Mutant> FilterMutants(IEnumerable<Mutant> mutants, FileLeaf file, StrykerOptions options) =>
```

richardwerkman commented on src/Stryker.Core/Stryker.Core/MutantFilters/ExcludeFromCodeCoverageFilter.cs:
```diff
@@ -0,0 +1,38 @@
+        IEnumerable<Mutant> IMutantFilter.FilterMutants(IEnumerable<Mutant> mutants, FileLeaf file, StrykerOptions options) =>
+            mutants.Where(m => !Exclude(m.Mutation.OriginalNode));
+
+        private static bool Exclude(SyntaxNode node) => 
```
Could you refactor these methods to regular methods? We don't use arrow methods like this anywhere else in the codebase.

riezebosch replied:
Sure thing, but not entirely true: https://github.com/stryker-mutator/stryker-net/blob/cee97b5113e4896df1443fadf8a31bd7514ab08d/src/Stryker.Core/Stryker.Core/MutantFilters/ExcludeMutationMutantFilter.cs#L17

richardwerkman replied:
Yeah true. But I think that is different from a whole expression that has to fit on another line 😉

richardwerkman commented on src/Stryker.Core/Stryker.Core/MutantFilters/ExcludeFromCodeCoverageFilter.cs:
```diff
@@ -0,0 +1,38 @@
+        private static bool HasExcludeFromCodeCoverageAttribute(MemberDeclarationSyntax m) =>
+            m.AttributeLists
+                .SelectMany(attr => attr.Attributes)
+                .Any(attr => Regex.IsMatch(attr.ToString(), @"^(?:System\.Diagnostics\.CodeAnalysis\.)?ExcludeFromCodeCoverage(?:Attribute)?$"));
```
I just learned that Namespace naming is enclosed in `QualifiedNameSyntax` nodes. So you could also get the `IdentifierNameSyntax` from the attribute DescendingNodesAndSelf(). We could get rid of the regex that way.

riezebosch replied:
Not sure how to fix this in the tests since I suspect it to not be available with the currently used code snippets.
